xml/schema/Core/Translation.xml

```diff
@@ -0,0 +1,122 @@
+    <comment>Unique String ID</comment>
+  </field>
+  <primaryKey>
+    <name>id</name>
+    <autoincrement>true</autoincrement>
+  </primaryKey>
+
+  <field>
+    <add>5.39</add>
+    <name>entity_table</name>
+    <title>Translated Entity</title>
+    <type>varchar</type>
+    <length>64</length>
+    <required>true</required>
+    <pseudoconstant>
+      <callback>CRM_Core_BAO_Translation::getEntityTables</callback>
+    </pseudoconstant>
+    <comment>Table where referenced item is stored</comment>
+  </field>
+
+  <field>
+    <add>5.39</add>
+    <name>entity_field</name>
+    <title>Translated Field</title>
+    <type>varchar</type>
+    <length>64</length>
+    <required>true</required>
+    <pseudoconstant>
+      <callback>CRM_Core_BAO_Translation::getEntityFields</callback>
+    </pseudoconstant>
+    <comment>Field where referenced item is stored</comment>
+  </field>
+
+  <field>
+    <add>5.39</add>
+    <name>entity_id</name>
+    <title>Translated Entity ID</title>
+    <type>int</type>
+    <length>64</length>
+    <required>true</required>
+    <comment>ID of the relevant entity.</comment>
+  </field>
+
+  <field>
+    <add>5.39</add>
+    <name>language</name>
+    <title>Language</title>
+    <type>varchar</type>
+    <length>5</length>
+    <required>true</required>
+    <comment>Relevant language</comment>
+    <html>
+      <type>Select</type>
+    </html>
+    <pseudoconstant>
+      <optionGroupName>languages</optionGroupName>
+      <keyColumn>name</keyColumn>
+      <optionEditPath>civicrm/admin/options/languages</optionEditPath>
+    </pseudoconstant>
+  </field>
+
+  <field>
+    <add>5.39</add>
+    <name>status_id</name>
+    <type>tinyint</type>
+    <title>Status</title>
+    <length>3</length>
+    <default>1</default>
+    <required>true</required>
+    <pseudoconstant>
+      <callback>CRM_Core_BAO_Translation::getStatuses</callback>
+    </pseudoconstant>
+    <comment>Specify whether the string is active, draft, etc</comment>
+  </field>
+
+  <field>
+    <add>5.39</add>
+    <title>Translated String</title>
+    <name>string</name>
+    <type>longtext</type>
+    <required>true</required>
+    <comment>Translated string</comment>
+  </field>
+
+  <dynamicForeignKey>
+    <add>5.39</add>
+    <idColumn>entity_id</idColumn>
+    <typeColumn>entity_table</typeColumn>
+  </dynamicForeignKey>
+
+  <index>
+    <add>5.39</add>
+    <!-- Expected queries:
+      "Admin UI: I'm editing a record. Show me all relevant translations."
+      "Public UI: I'm browsing a list of records. Show this page-worth of records in my preferred language."
+    -->
+    <name>index_entity_lang</name>
+    <!-- Prediction: In a large DB with many events/contribution-pages/groups/mailings/etc, entity ID will have best selectivity. -->
+    <!-- Prediction: Over diverse set of deployments, the selectivity of 'table' and 'language' will be similar. -->
+    <fieldName>entity_id</fieldName>
+    <fieldName>entity_table</fieldName>
```

Comment:
@totten note that in a combined index the second 2 parts of the index will never be hit if the first is not - ie
```
SELECT * FROM civicrm_translation WHERE id = 5 AND entity_table = 'civicrm_discount' AND language = 'en_US'
```
will utilise the full index.

If the query is 

```
SELECT * FROM civicrm_translation WHERE entity_table = 'civicrm_discount' AND language = 'en_US'
```
then the index will not be used at all.

For this query


```
SELECT * FROM civicrm_translation WHERE id = 5 AND AND language = 'en_US'
```

the index on id is used (yay) but not on language


I'm actually fine merging as is - I suspect the entity_id is what we really need but it might be over time that we add other other indexes on the other 2 fields so we can retrieve all by language or all by entity.

Reply:
> I'm actually fine merging as is - I suspect the entity_id is what we really need but it might be over time that we add other other indexes...

Yeah, this is my theory too - at least for the current cases where you start by filtering the main entity (`MessageTemplate`, `Event`, etc) and then (as a final step) merge-in the translations for the output.

When/if we have a case for textual filtering on translated columns, then we'll need to consider other indexes...